blocks2image: modal instead of menu items & new "copy to clipboard"

Jazza-231 opened the pull request:
Resolves #6823 and also my own suggestion
Resolves #6096

### Changes

Lets you right click to copy the image of blocks to your clipboard

![image](https://github.com/ScratchAddons/ScratchAddons/assets/17484114/50ed1b6b-3ad1-457f-926b-302f5a318673)

![image](https://github.com/ScratchAddons/ScratchAddons/assets/17484114/3a700da6-c97f-4ed1-ac9d-909696e491b1)

### Reason for changes

Convenience

### Tests

Tested on chrome and unlike the embarassment of my last pr im pretty sure there are no problems

Jazza-231 commented:
https://github.com/ScratchAddons/ScratchAddons/issues/6823#issuecomment-1786751255

mybearworld commented:
> [#6823 (comment)](https://github.com/ScratchAddons/ScratchAddons/issues/6823#issuecomment-1786751255)

Maybe you could add an "as PNG/SVG" to the end of the options?

Jazza-231 commented:
any word on this?

mxmou commented on addons/blocks2image/addon.json:
```diff
@@ -17,5 +17,10 @@
-  "enabledByDefault": false
+  "enabledByDefault": false,
+  "latestUpdate": {
+    "version": "1.35.0",
+    "temporaryNotice": "You can now copy directly to the system clipboard"
```
```suggestion
    "version": "1.36.0",
    "temporaryNotice": "You can now copy directly to the system clipboard."
```

mxmou commented on addons/blocks2image/userscript.js:
```diff
@@ -270,9 +286,19 @@ export default async function ({ addon, console, msg }) {
+          .copyImage(dataURL)
+          .then(() => console.log("Image successfully copied"))
+          .catch((e) => console.error(`Image could not be copied: ${e}`));
+      }
+
+      if (!copy) {
```
```suggestion
      } else {
```

WorldLanguages commented:
> any word on this?

I'm not convinced about adding so many items to the right click menu, this is also related to #4919

DNin01 commented:
> > any word on this?
> 
> I'm not convinced about adding so many items to the right click menu, this is also related to #4919

We should bundle all of the script snapshot-related options in a dropdown menu if possible.

Jazza-231 commented:
> > > any word on this?
> > 
> > 
> > I'm not convinced about adding so many items to the right click menu, this is also related to #4919
> 
> We should bundle all of the script snapshot-related options in a dropdown menu if possible.

A dropdown menu in the right click menu?

DNin01 commented:
> > > > any word on this?
> > > 
> > > 
> > > I'm not convinced about adding so many items to the right click menu, this is also related to #4919
> > 
> > 
> > We should bundle all of the script snapshot-related options in a dropdown menu if possible.
> 
> A dropdown menu in the right click menu?

Yes, that's what I meant.

Jazza-231 commented:
im not going to mess with the API at this stage so if anyone feels the need to implement this, go ahead lol

Samq64 commented on addons/blocks2image/addon.json:
```diff
@@ -17,5 +17,10 @@
+    "version": "1.36.0",
+    "temporaryNotice": "You can now copy directly to the system clipboard."
+  },
+  "permissions": ["clipboardWrite"]
```
In the future it could interesting to offer per-setting permissions and put the clipboard options behind a toggle.

Samq64 commented on addons/blocks2image/userscript.js:
```diff
@@ -90,6 +90,14 @@ export default async function ({ addon, console, msg }) {
+          enabled: true,
+          text: msg("copy_all_to_clipboard"),
+          callback: () => {
+            exportBlock(true, "", true);
```
This isn't a big deal but  the empty argument could be removed if the 2nd and 3rd were swapped everywhere.

Samq64 commented on addons/blocks2image/userscript.js:
```diff
@@ -270,9 +286,17 @@ export default async function ({ addon, console, msg }) {
+      if (copy) {
+        addon.tab
+          .copyImage(dataURL)
+          .then(() => console.log("Image successfully copied"))
```
This is probably unnecessary?

Samq64 commented on addons-l10n/en/blocks2image.json:
```diff
@@ -3,5 +3,7 @@
   "blocks2image/export_selected_to_PNG": "Export block as PNG",
   "blocks2image/export_all_to_PNG": "Export all as PNG",
-  "blocks2image/error_blocks_not_added": "Add blocks to workspace!"
+  "blocks2image/error_blocks_not_added": "Add blocks to workspace!",
```
This really doesn't matter but this at the bottom to keep all the context menu strings together.

Samq64 reviewed (approved):
I have a few minor nitpicks but everything works fine.

Samq64 commented:
An alternative to the submenu could be to have settings to turn of some of the options, but it's just 2 more items.

Samq64 commented:
Could the description also mention copying?

Jazza-231 commented:
> Could the description also mention copying?

Sure, do you have a suggestion?

Samq64 commented:
> > Could the description also mention copying?
> 
> Sure, do you have a suggestion?

How about "Right click the code area to copy blocks to clipboard or download them as an SVG or PNG."?

WorldLanguages commented:
> An alternative to the submenu could be to have settings to turn of some of the options, but it's just 2 more items.

Maybe we could hide this by default (and make the 2 new buttons a settings) until we can figure out how to handle submenus?

Samq64 commented:
> I think other ways to preview block screenshots were suggested somewhere else in the repository

Is it #6632? There's also the copy BBCode suggestion: #2336.

Samq64 commented:
> What just happened here

I think I accidentally did a full non-fast forward merge with upstream/master. it's fixed now.

Jazza-231 commented:
Hey, trying to get some of these things finished since I have a rare break, can someone tell me what, if anything, they want me to do to get this merged?

Samq64 commented:
> Hey, trying to get some of these things finished since I have a rare break, can someone tell me what, if anything, they want me to do to get this merged?

This just needs a second approval.

DNin01 commented:
```
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'cloneNode')
```

Jazza-231 commented:
> It just throws an uncaught exception at line 169. Probably because the arguments are swapped from what they should be:
> 
> ```js
> async function exportBlock(isExportPNG, block, copyToClipboard) {
> ```
> 
> ```js
> exportBlock(false, false, block); // Should be false, block, false
> ```

You seem to be using an old version of this. 
`  function exportBlock(isExportPNG, copyToClipboard, block)`
this is what i have
![image](https://github.com/ScratchAddons/ScratchAddons/assets/81956724/ace3403f-8c54-4db3-a26f-16ce61efd881)

Jazza-231 commented:
@Samq64 i think your little accident has fucked things up

Jazza-231 commented:
Besides the dropdown menu in the right click menu this is ok. Do people really want that? if yes then uhmmmmm im not sure how to go about that, if someone wants to do it themself or go through the pain of giving me some quick instruction or a todo list for how to implement it then that would be great

DNin01 reviewed (approved):
Everything works AFAIK.

TheColaber commented:
These changes work, but maybe we should consider my comment in the issue:
> I had an idea to rework this entire addon, to help with these changes but to also make it take up less options in the right click menu. So instead of there being those 4 options when you right click on the block, there would be one: "Save block as image". When you click it, it will open up a popup (like how we do for the recorder addon) showing what block/stack was selected. From there, you can select to just save the block or the entire stack, and if you want to download as an svg, png, or save to clipboard. It would be 3 clicks (Right click block, Select "Save block as image", Click "Copy to clipboard"), while at the same time introducing better UI/UX and taking up 3 less options in the right click menu.

This would take up 5 less options now if this PR were merged. We should consider this course of action. I don't prefer submenus because they add an extra layer of complexity. A popup is a better UX for the user.

Jazza-231 commented:
> These changes work, but maybe we should consider my comment in the issue:
> 
> > I had an idea to rework this entire addon, to help with these changes but to also make it take up less options in the right click menu. So instead of there being those 4 options when you right click on the block, there would be one: "Save block as image". When you click it, it will open up a popup (like how we do for the recorder addon) showing what block/stack was selected. From there, you can select to just save the block or the entire stack, and if you want to download as an svg, png, or save to clipboard. It would be 3 clicks (Right click block, Select "Save block as image", Click "Copy to clipboard"), while at the same time introducing better UI/UX and taking up 3 less options in the right click menu.
> 
> This would take up 5 less options now if this PR were merged. We should consider this course of action. I don't prefer submenus because they add an extra layer of complexity. A popup is a better UX for the user.

Huh I actually really like that, its a cool idea, I would totally make that except for the showing what block/s were selected, it think I would need help with that part (probably from one of the gods/goddesses of blockly like garbo or taco)

TheColaber commented:
> I would totally make that except for the showing what block/s were selected, it think I would need help with that part 

@Jazza-231 just use the svg/png and put it in an image tag

Jazza-231 commented:
> > I would totally make that except for the showing what block/s were selected, it think I would need help with that part
> 
> @Jazza-231 just use the svg/png (that this addon already generates) and put it in an image tag

My lord do I overcomplicate things

Jazza-231 commented:
![1 pixel off](https://github.com/ScratchAddons/ScratchAddons/assets/81956724/07cd6c60-90ab-4387-b6c5-ca8a1b4b7fdd)
For context, this is the fix that last commit did.

TheColaber reviewed (commented):
love it

Jazza-231 commented on addons/blocks2image/userscript.js:
```diff
@@ -87,33 +75,139 @@ export default async function ({ addon, console, msg }) {
   );
 
-  async function exportBlock(isExportPNG, block) {
+  function handleNumberkey(key) {
+    switch (key) {
+      case "1":
+        document.querySelector(".sa-export-copy-button").click();
+        break;
+      case "2":
+        document.querySelector(".sa-export-svg-button").click();
+        break;
+      case "3":
+        document.querySelector(".sa-export-png-button").click();
+        break;
+    }
+  }
```
The original design had a longer function for each of the buttons making code repetition a big deal so I chose this method, but it is no longer applicable so I will remove it. Good catch.

Jazza-231 commented on addons/blocks2image/addon.json:
```diff
@@ -1,11 +1,15 @@
 {
   "name": "Save blocks as image",
-  "description": "Right click the code area to export blocks as SVG/PNG images.",
+  "description": "Right-click on the code area or a block/stack of blocks to copy it to the clipboard or save it as a PNG or SVG file.",
```
Yea ok

Jazza-231 commented:
I removed the +18 from `allBlocks` function, to go from this result:

![image](https://github.com/ScratchAddons/ScratchAddons/assets/81956724/0ee02bae-626a-435d-a009-5afb35ab3258)


To this:

![image](https://github.com/ScratchAddons/ScratchAddons/assets/81956724/24fc1ae0-e983-4ad3-a2c9-a9af0eee5e31)

Not sure why this happens but yeah. Now there's just the problem of the one missing pixel around every edge.

DNin01 commented:
In-between a modal and too many context menu options would be:

We have a single button to save a block as an image, but instead of opening a menu, it opens a second context menu. We could do this by removing all the options and then adding options for saving or copying the block. We'd also have to prevent the menu from automatically closing when the parent option is chosen.

Implementing a sub-selection into the context menu may sound daunting, but we can still try to get clever. I think it's a better user experience too.

Jazza-231 commented:
> In-between a modal and too many context menu options would be:
> 
> We have a single button to save a block as an image, but instead of opening a menu, it opens a second context menu. We could do this by removing all the options and then adding options for saving or copying the block. We'd also have to prevent the menu from automatically closing when the parent option is chosen.
> 
> Implementing a sub-selection into the context menu may sound daunting, but we can still try to get clever. I think it's a better user experience too.

I prefer the suggested idea of having a drop down menu, where instead of replacing the current context menu with completely new items as you suggest, just making a new drop down context menu that shows to the right. 

But above all, the modal is my favourite choice, seeing the blocks is actually really helpful for understanding what is being saved.

DNin01 commented:
> But above all, the modal is my favourite choice, seeing the blocks is actually really helpful for understanding what is being saved.

Hmm… does this call for an addon setting?

Jazza-231 commented:
> > But above all, the modal is my favourite choice, seeing the blocks is actually really helpful for understanding what is being saved.
> 
> Hmm… does this call for an addon setting?

Please no 😭I want this to be overrrrr

TheColaber commented:
I think the model makes the most sense here, since usually users wont be downloading/coping blocks to images very often. If it was often, it would make sense for it to be a more accessible context menu option. I think this addon is good as is.

WorldLanguages commented:
> I removed the +18 from `allBlocks` function, to go from this result:

If possible could this be moved to a separate PR? Not very far ago we merged #5604 which also changed them, and it's quite easy to break things.

WorldLanguages commented:
Ideally I think we should default to 1 context menu option which opens the modal.

Maybe, after the first image is exported within a session, we can add a second context menu item that saves the user a click. It could be useful if exporting lots of blocks within a short period of time. We would assume that if the first time they exported as PNG, they want an export as PNG option in the context menu, etc.

DNin01 commented:
> we can add a second context menu item that saves the user a click.

Would it just be the same as the last option the user chose from the modal? So, for example, if the user chose to save an SVG of some blocks, the shortcut option in the context menu would be to save as SVG?

WorldLanguages commented:
@DNin01 Yes, that's what I am thinking. I'm not proposing it should be done within this PR, though.

Jazza-231 commented:
> > I removed the +18 from `allBlocks` function, to go from this result:
> 
> If possible could this be moved to a separate PR? Not very far ago we merged #5604 which also changed them, and it's quite easy to break things.

Just to make my life easier I would rather not, I've tested it very extensively and it all works.



> @DNin01 Yes, that's what I am thinking. I'm not proposing it should be done within this PR, though.

That doesn't sound like a bad idea, but considering that saves...one click or button press (you can click and press the number keys REALLY quickly) it's not needed

WorldLanguages commented:
@Jazza-231 Is the "save all as an image" feature also new?

WorldLanguages commented on addons/blocks2image/addon.json:
```diff
@@ -16,6 +20,20 @@
+      "text": "You can press the enter key to copy the block to the clipbard, or press the 1 - 3 keys corresponding to the buttons."
+    }
+  ],
+  "permissions": ["clipboardWrite"]
```
This may disable the addon for all Firefox users, even if they don't plan to use the clipboard feature. Please don't close this comment. I'm researching how we can proceed

Samq64 replied:
I mentioned per-setting permissions earlier, but I don't think any other addons would use it (https://github.com/ScratchAddons/ScratchAddons/pull/6830#discussion_r1451782282).

WorldLanguages replied:
In the latest version of Firefox, the clipboardWrite permission can be ignored (it is already ignored in Chrome). See #7625

DNin01 replied:
We could make it so that Scratch Addons doesn't respond to missing permissions (that aren't crucial) by disabling addons, or we could request the necessary permissions somehow.

WorldLanguages replied:
I created PR #7630.

Only around ~1000 users run Firefox versions lower than 127, and of course not all of them use this addon, so it's only a few users that will see this addon turn off automatically.
This addon will be positioned at the top of the settings page due to its "new" status, so it will be easy to find it.

WorldLanguages replied:
Solved by PR #7630

Jazza-231 commented:
> @Jazza-231 Is the "save all as an image" feature also new?

No

sfederici commented:
> For this PR, my suggestion is that there should only be 1 context menu and there is no good reason to add settings to this addon.

As I would use it a lot, I would love having settings that would allow me to choose between direct copy in the clipboard or dialog

WorldLanguages commented:
> You can press the enter key to copy the block to the clipbard, or press the 1 - 3 keys corresponding to the buttons.

I'm going to remove this feature for now. For the next release, v1.40, we can react to feedback (including existing feedback by sfederici) on the best way to go to save users one or two clicks when exporting multiple blocks.

WorldLanguages commented:
> > > > I removed the +18 from `allBlocks` function, to go from this result:
> > 
> > 
> > If possible could this be moved to a separate PR? Not very far ago we merged #5604 which also changed them, and it's quite easy to break things.
> 
> Just to make my life easier I would rather not, I've tested it very extensively and it all works.

I'm also going to move "image logic" to another PR, but will merge the two simultaneously. The benefit is that we can revert those changes separately if a hotfix is necessary.

DNin01 commented:
A modal would be a good opportunity to have even more export options, such as additional image formats, resolution/quality options, or whether to apply Scratch Addons' styles or export them with the Scratch original style and which one of Scratch's color modes to use.

WorldLanguages commented:
This is blocked by PR #7630 but we will sort it out quickly.

WorldLanguages commented:
Okay, the blocking PR was merged.

WorldLanguages commented on addons-l10n/en/blocks2image.json:
```diff
@@ -1,7 +1,8 @@
+  "blocks2image/save": "Save as an image",
+  "blocks2image/modalTitle": "Blocks to image",
+  "blocks2image/clipboard": "Copy to clipboard",
+  "blocks2image/svg": "Download as a SVG",
+  "blocks2image/png": "Download as a PNG"
```
Sounds weird to say "a SVG", I may change it...

```suggestion
  "blocks2image/svg": "Download as SVG",
  "blocks2image/png": "Download as PNG"
```

Hans5958 replied:
Would be more appropriate to do "an SVG" ("an es-vee-gee") grammar-wise, but this is better.

WorldLanguages commented on addons/blocks2image/addon.json:
```diff
@@ -1,11 +1,15 @@
 {
   "name": "Save blocks as image",
-  "description": "Right click the code area to export blocks as SVG/PNG images.",
+  "description": "Right-click on the code area or a stack of blocks to copy it to the clipboard or save it as a PNG or SVG file.",
```
Consistency with other addon descriptions.
```suggestion
  "description": "Right click on the code area or a stack of blocks to copy it to the clipboard or save it as a PNG or SVG file.",
```

WorldLanguages replied:
The manifest docs say to use `scripts` instead of `stacks of blocks`, but eh... I think "stack of blocks" is more clear here, and 3 other addons already omit this rule.

WorldLanguages commented on addons/blocks2image/addon.json:
```diff
@@ -1,11 +1,15 @@
+    {
+      "name": "Jazza",
+      "link": "https://scratch.mit.edu/users/greeny--231"
     }
```
Can also add mxmou if he wants

```suggestion
    },
    {
      "name": "World_Languages"
    }
```

DNin01 reviewed (commented):
I think the blocks are a bit big on the modal.

Jazza-231 commented:
> I think the blocks are a bit big on the modal.

Feel free to give suggestions

sfederici commented:
> > I think the blocks are a bit big on the modal.
> 
> Feel free to give suggestions

I guess the maximum size of the blocks could be the actual size of the blocks in the IDE

DNin01 commented:
> > > I think the blocks are a bit big on the modal.
> 
> > 
> 
> > Feel free to give suggestions
> 
> 
> 
> I guess the maximum size of the blocks could be the actual size of the blocks in the IDE

…Give or take a little.

WorldLanguages commented:
Second review anyone?

Samq64 commented:
For large images the modal is just the header and the buttons require scrolling to access. Also there's twice as much loading time overall since the image is re-generated after actually clicking the button.

Jazza-231 commented:
> For large images the modal is just the header while loading

Yeah I'll fix that
> buttons require scrolling to access after it does load

I can't repro this?
> Also there's twice as much loading time overall since the image is re-generated after actually clicking the button.
Can't do anything about this, if we choose to generate it as an image and the user saves it as an SVG, no luck, if we choose to generate it as an SVG and the user copies it as a png, no luck

WorldLanguages commented:
> buttons require scrolling to access after it does load

@Samq64 Could you share a screenshot?

Samq64 commented:
![image](https://github.com/user-attachments/assets/aa2668c2-633c-4def-9b9f-cf5d5a5476ea)

Samq64 commented:
> > For large images the modal is just the header while loading
> 
> Yeah I'll fix that (what do you suggest)

Just some text as the main content. Maybe "Loading preview..."?

Samq64 commented:
Ideally the preview should be async so the buttons still work.

WorldLanguages commented:
> Ideally the preview should be async so the buttons still work.

Can we know the dimensions of the image before rendering it? Probably not.

How many seconds does it take in your device to render the preview? Is it significant?

Samq64 commented:
> How many seconds does it take in your device to render the preview? Is it significant?

3-ish seconds for 166 blocks (Steve arm sprite in paper Minecraft)

WorldLanguages commented:
I don't think people will use this often to export so many blocks.

Jazza-231 commented:
This is some interesting information from you....

The steve arm sprite takes about 400ms for me to load the preview, and does not hide the buttons
![image](https://github.com/user-attachments/assets/0d96e66d-b18f-4fd3-9568-a56a01c5da48)

Jazza-231 commented:
The biggest issue in my eyes is the fact that the image generation flat out fails on too many blocks (or too wide of an image), for example try exporting the blocks in the `steve` sprite from [paper minecraft](https://scratch.mit.edu/projects/10128407/editor/)

![image](https://github.com/user-attachments/assets/f14e421d-6f1b-4c76-8004-4385aacd76e8)

Samq64 commented:
> The steve arm sprite takes about 400ms (edit: 250ms) for me to load the preview

Do you happen to have `editor-theme3` enabled? For me it seems to significantly improve the performance for whatever reason.

WorldLanguages commented:
@Samq64 That's curious... Enabling editor-theme3 speeds up things by 10x? (Maybe it already did before this PR's changes, though!)

Samq64 commented:
> Maybe it already did before this PR's changes, though!

That seems to be the case, so it can be ignored in this PR.

WorldLanguages commented:
I think this is OK to merge and we can fix #7681 separately, it is a bug that predates this PR. What do you think?

Samq64 commented:
> I think this is OK to merge and we can fix #7681 separately, it is a bug that predates this PR. What do you think?

I can still reproduce the scrollbar bug with the Steve arm sprite with the browser maximized on a 1440p monitor.

WorldLanguages commented:
Yeah, ideally the export buttons should always be visible, even if the preview can be scrolled vertically.

(Sprite "Steve arm" on https://scratch.mit.edu/projects/10128407/editor/)

![image](https://github.com/user-attachments/assets/5836aff2-077c-4178-ab8b-cd2d29a89227)

Jazza-231 commented:
> I can still reproduce the scrollbar bug with the Steve arm sprite and the browser maximized on a 1440p monitor.


I have a 1600p screen so maybe that small difference is enough. I'm pretty sure there is css I can do to make sure the buttons stay at the bottom no matter what, I thought I did but I'll look at it another time. Can't rn, so @WorldLanguages please remind me on discord in about 9 hours.

WorldLanguages commented:
@mxmou If you have some time maybe you can help with the CSS of the modal to avoid the bug I mentioned yesterday.

mxmou commented:
> @mxmou If you have some time maybe you can help with the CSS of the modal to avoid the bug I mentioned yesterday.

I can't test this right now, but I think making `.sa-export-content` a flex container and adding `flex-basis: 0` and `flex-grow: 1` to the image container would work.

Jazza-231 commented:
On the modal header:
    position: fixed;
    width: 692px;
    border-radius: inherit;

I have no idea why this works so its a bad idea. But it does work, so its worth sending here? Mxmou's suggestion doesn't work without a min-width, and doesn't work well with it. I am not sure why there is not a simple solution to this? This is essentially a footer, yet it seems to not work the same.

mxmou commented:
> But it does work, so its worth sending here?

You can push the code, then I'll try to improve it when I have time.

sfederici commented:
Can someone please approve this update so that it will be available in the next release? Thanks in advance.

WorldLanguages commented:
@sfederici No, this is planned for v1.40
For personal use you may download this branch [as a zip file](https://github.com/Jazza-231/ScratchAddons/archive/refs/heads/copy-blocks-to-clip.zip) and manually add it to your browser by enabling developer mode in `chrome://extensions`.

Samq64 commented on addons/blocks2image/style.css:
```diff
@@ -0,0 +1,57 @@
+.sa-export-container {
+  box-sizing: border-box;
+  width: 700px;
+  max-height: min(800px, 80vh);
```
```suggestion
```
Removing this seems to fix the double scrollbar issue.

Secret-chest commented:
How do I copy SVG to clipboard?

Jazza-231 commented:
> How do I copy SVG to clipboard?

This is not useful enough to be added, as of now you can not copy the svg to your clipboard. The clipboard is not designed for that amount of data in that format (essentially text).
